remove deprecated `std::result_of`

`src/tasks/tasks.hpp` uses `std::result_of`, which is deprecated in C++17 and removed in C++20. This causes lots of warnings on my machine:

```
In file included from /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:30:
/Users/benwibking/athenapk/external/parthenon/src/tasks/thread_pool.hpp:121:36: warning: 'result_of<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20) ()>' is deprecated [-Wdeprecated-declarations]
    using return_t = typename std::result_of<F(Args...)>::type;
                                   ^
/Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:12: note: in instantiation of function template specialization 'parthenon::ThreadPool::enqueue<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20)>' requested here
      pool.enqueue([t, &ProcessTask]() { return ProcessTask(t); });
           ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__type_traits/result_of.h:24:34: note: 'result_of<(lambda at /Users/benwibking/athenapk/external/parthenon/src/tasks/tasks.hpp:420:20) ()>' has been explicitly marked deprecated here
template <class _Callable> class _LIBCPP_DEPRECATED_IN_CXX17 result_of;
                                 ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:862:41: note: expanded from macro '_LIBCPP_DEPRECATED_IN_CXX17'
#    define _LIBCPP_DEPRECATED_IN_CXX17 _LIBCPP_DEPRECATED
                                        ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.2.sdk/usr/include/c++/v1/__config:835:49: note: expanded from macro '_LIBCPP_DEPRECATED'
#      define _LIBCPP_DEPRECATED __attribute__((deprecated))
                                                ^
4 warnings generated.
```

--- a/src/tasks/thread_pool.hpp
+++ b/src/tasks/thread_pool.hpp
@@ -151,7 +151,7 @@ class ThreadPool {
 
   template <typename F, class... Args>
   void enqueue(F &&f, Args &&...args) {
-    using return_t = typename std::result_of<F(Args...)>::type;
+    using return_t = typename std::invoke_result_t<F, Args...>;
     auto task = std::make_shared<std::packaged_task<return_t()>>(
         [=, func = std::forward<F>(f)] { return func(std::forward<Args>(args)...); });
     // If we're listing Prathenon "Tasks" (all current uses) keep the returns

